rust/src/nasl/builtin/ssh/russh/session.rs

```diff
@@ -0,0 +1,235 @@
+    async fn disconnected(
+        &mut self,
+        reason: DisconnectReason<Self::Error>,
+    ) -> Result<(), Self::Error> {
+        match reason {
+            DisconnectReason::ReceivedDisconnect(_) => Ok(()),
+            DisconnectReason::Error(e) => {
+                match e {
+                    russh::Error::Disconnect => {}
+                    _ => {
+                        error!("SSH session disconnected due to error: {}", e);
+                    }
+                }
+                Err(e)
+            }
+        }
+    }
+}
+
+/// This struct is a convenience wrapper
+/// around a russh client
+pub struct SshSession {
+    id: SessionId,
+    session: client::Handle<Client>,
+}
+
+impl SshSession {
+    // The alternative to many arguments here is making
+    // the fields public or creating some sort of builder struct
+    // both of which feel like much worse alternatives.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn new(
+        id: SessionId,
+        ip_addr: IpAddr,
+        port: Port,
+        timeout: Option<Duration>,
+        keytype: Vec<key::Name>,
+        csciphers: Vec<cipher::Name>,
+        scciphers: Vec<cipher::Name>,
+        socket: Option<Socket>,
+    ) -> Result<Self, SshError> {
+        if socket.is_some() {
+            error!("Using custom sockets not yet implemented.");
+            return Err(SshErrorKind::Unimplemented.with(id));
+        }
+        let preferred = construct_preferred(keytype, csciphers, scciphers);
+        let config = client::Config {
+            inactivity_timeout: timeout,
+            preferred,
+            ..Default::default()
+        };
+
+        let config = Arc::new(config);
+        let sh = Client {};
+
+        let session = connect(config, (ip_addr, port), sh)
+            .await
+            .map_err(|e| SshErrorKind::Connect.with(id).with(e))?;
+
+        Ok(Self { session, id })
+    }
+
+    pub async fn exec_ssh_cmd(&self, command: &str) -> Result<(String, String), SshError> {
+        let stdout = self.call(command).await.map_err(|e| {
+            SshErrorKind::RequestExec(command.to_string())
+                .with(self.id)
+                .with(e)
+        })?;
+        // TODO implement stderr properly.
+        let stderr = String::new();
+        Ok((stdout, stderr))
+    }
+
+    pub async fn call(&self, command: &str) -> Result<String, russh::Error> {
+        let mut channel = self.session.channel_open_session().await?;
+        channel.exec(true, command).await?;
+
+        let mut code = None;
+        let mut stdout = String::new();
+
+        loop {
+            // There's an event available on the session channel
+            let Some(msg) = channel.wait().await else {
+                break;
+            };
+            match msg {
+                // Write data to the terminal
+                ChannelMsg::Data { ref data } => {
+                    stdout.push_str(&bytes_to_str(data));
```

Comment:
q: did you verify that this is just stdout? what happens with stderr data?

Reply:
I had a todo a few lines above because I hadn't implemented stderr properly, but now I implemented stderr for russh (turns out that for stderr, you receive an `ExtendedData` variant with a specific number set to exactly 1. Love the SSH protocol)  and added a "test" for it.

I also added a comment explaining that the compatibility mode for `ssh_request_exec` is currently not implemented as described (and also wasn't before this PR). We should check whether this is ever used in the described way and then either get rid of it or implement it properly.